docs/dev/specs/packable.md

```diff
@@ -0,0 +1,284 @@
+}
+```
+
+- The `Error` associated type represents a semantic error.
+
+- The `pack` method serializes the current value using a `Packer` to write the
+  bytes.
+
+- The `unpack` method deserializes a value using an `Unpacker` to read the
+  bytes.
+
+- The `packed_len` method returns the length in bytes of the serialized value.
+  This has to match the number of bytes written using `pack` to avoid
+  inconsistencies.
+
+## Error handling
+
+In addition to the three `Error` associated types mentioned before, we provide
+two new helper error types:
+
+- The `UnpackError` type which is an enum wrapping values of either the
+  `Unpacker::Error` or `Packable::Error` type.
+
+- The `UnknownTagError` type which can be used if the prefix tag used to
+  represent the variant of an enum does not correspond to any variant of such
+  enum. As a convention, any `Packable::Error` type for an enum must implement
+  `From<UnknownTagError>` (this is not enforced in the trait itself because
+  this error is not used for structs).
+
+We also use `core::convert::Infallible` as `Packable::Error` for types whose
+unpacking does not have semantic errors, like integers for example.
+
+# Usage
+
+## An example
+
+### Packing
+
+We will implement `Packable` for a type that encapsulates optional integer values (like
+`Option<i32>`):
+
+```rust
+pub enum Maybe {
+    Nothing,
+    Just(i32),
+}
+```
+Following the conventions from the [IOTA protocol messages RFC](https://github.com/iotaledger/protocol-rfcs/pull/0017),
+we will use an integer prefix as a tag to determine which variant of the enum
+is being packed
+
+```rust
+use bee_common::packable::{Packable, Packer, UnknownTagError, Unpacker, UnpackError};
+
+impl Packable for Maybe {
+    type Error = UnknownTagError<u8>;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        match self {
+            Self::Nothing => 0u8.pack(packer),
+            Self::Just(value) => {
+                1u8.pack(packer)?;
+                value.pack(packer)
+            },
+        }
+    }
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+        match u8::unpack(unpacker).map_err(UnpackError::coerce)? {
+            0u8 => Ok(Self::Nothing),
+            1u8 => Ok(Self::Just(i32::unpack(unpacker).map_err(UnpackError::coerce)?)),
+            tag => Err(UnpackError::Packable(UnknownTagError(tag))),
+        }
+    }
+
+    fn packed_len(&self) -> usize {
+        match self {
+            Self::Nothing => 0u8.packed_len(),
+            Self::Just(value) => 1u8.packed_len() + value.packed_len(),
+        }
+    }
+}
+```
+
+
+## The `derive` macro
+
+Another option is to use the derive macro for the `Packable` trait which
+provides an implementation equivalent to the one written before.
+
+```rust
+use bee_common::packable::{Packable, Packer, UnknownTagError, Unpacker, UnpackError};
+
+#[derive(Packable)]
+#[packable(tag_ty = u8)]
```

Comment:
Looks like this is actually `tag_type` (at least, that's what the compiler is telling me!)

Reply:
yes it is, it used to be `tag_Ty` a few versions back.